MySQL tracing without USDT

WIgor opened the pull request:
What do you think about tracing MySQL without recompilation with USDT support? 

Actually, there is some downside, at least temporary, this script supports only 5.6 version and derivatives (MariaDB, Percona Sever 5.6, etc.), may be lower version also - but it wasn't tested.

From another side, starting from 8.0 MySQL, as I understand documentation, DTrace tracepoints were removed.

This script isn't a final version, just example for discussion.

goldshtn commented:
Hi Igor, it's definitley a good idea. In fact, it can be a good idea for any runtime or product that doesn't have USDT probes enabled by default -- which is, unfortunately, many of them. As you noted, using uprobes for this is a bit fragile -- functions are prone to moving and changing without notice -- but it's the next best thing if there is no static tracing available.

If you want to follow this through, it might be nice to actually consolidate this into the existing dbslower tool, so that it tries to use USDT probes and if they are not available, falls back to uprobes. How's that sound?

WIgor commented:
Yes. A movement to dbslower tool looks good, a code is almost same.

WIgor commented:
I updated dbslower, added support for MySQL 5.7. 

But for PostgreSQL same functionality might be not easy at all. Most interesting function (that contains tracing probes) exec_simple_query - is static.

goldshtn commented on tools/dbslower.py:
```diff
@@ -101,17 +174,61 @@
+    # Uprobes mode
+    program = program_uprobe
+
+    symbols = subprocess.check_output(["nm", "-aD", args.path])
```
We have built-in symbol support in the BPF module. Can you see if that can be used here, instead of invoking nm directly?

WIgor replied:
Can you point me an example? Initially, I tried to find more "pythonian" way to do this, but haven't found yet.

goldshtn replied:
I think you want [`BPF.get_user_functions_and_addresses`](https://github.com/iovisor/bcc/blob/master/src/python/bcc/__init__.py#L770). It's not Pythonian, it's a wrapper on top of libbcc, which uses libelf for this.

goldshtn commented on tools/dbslower.py:
```diff
@@ -39,21 +46,87 @@
+    u64 pid = bpf_get_current_pid_tgid();
+    u64 command  = (u64) PT_REGS_PARM1(ctx);
+
+    if (command == 3) {
```
It might be nice to add some documentation around these magic numbers.

goldshtn commented on tools/dbslower.py:
```diff
@@ -39,21 +46,87 @@
+
+"""
+
+program_udst = """
```
There's quite a bit of code duplication, especially here around the result reporting. Perhaps instead of having two completely separate programs, you could use conditional compilation or text replacement to have as much shared code as possible? When the scripts grow, any duplication becomes a maintenance burden.

goldshtn commented on tools/dbslower.py:
```diff
@@ -2,14 +2,19 @@
 # dbslower      Trace MySQL and PostgreSQL queries slower than a threshold.
 #
-# USAGE: dbslower [-v] [-p PID [PID ...]] [-m THRESHOLD] {mysql,postgres}
+# USAGE: dbslower [-v] [-p PID [PID ...]] [-b PATH_TO_BINARY] [-m THRESHOLD] {mysql,postgres}
```
I think I'd go with `-x` for executable, or `-l` for library, and not `-b`, which I think we used for buffer sizes in other tools. It's not very important, but it's nice to have consistency across different tools.

goldshtn commented on tools/dbslower.py:
```diff
@@ -128,8 +245,8 @@ def print_event(cpu, data, size):
 
-print("Tracing database queries for pids %s slower than %d ms..." %
-      (', '.join(map(str, args.pids)), args.threshold))
+# print("Tracing database queries for pids %s slower than %d ms..." %
```
Was this intentionally commented out?

WIgor replied:
Accidentally, forgot to update this line.

goldshtn commented on tools/dbslower.py:
```diff
@@ -101,17 +174,61 @@
+        if m:
+            func_name = m.group(0)
+        else:
+            print("Cant extract real 'dispatch_command' function name from %s" % (dispatch_fname[0]))
```
This should be "Can't".

WIgor commented:
I updated script according to comments. Tested it in uprobe modes, but haven't tested it with USDT, yet. As I understand - USDT works only with Oracle Linux or Solaris.

goldshtn commented:
USDT is available on Linux. You need to build MySQL with the appropriate configure flag to enable DTrace probes.

WIgor commented:
Am I right that for this I need https://github.com/dtrace4linux/linux?

goldshtn commented:
No. Just build MySQL with the right configure switch. I'm on mobile right now but it should be googleable. You'd need the systemtap-sdt-devel package or similar for your distri.

WIgor commented:
cmake ../mysql-server -DENABLE_DTRACE=1 -DDOWNLOAD_BOOST=1 -DWITH_BOOST=~/boost

Gives an error during next make.

goldshtn commented:
Which distribution are you building on? You need the systemtap-sdt-dev package on Ubuntu, or the systemtap-sdt-devel package on Fedora, to get the dtrace command-line tool.

WIgor commented:
Yes, with systemtap-sdt-dev, compiles. And USDT mode works.

goldshtn commented:
[buildbot, ok to test]

goldshtn commented:
@WIgor Thanks for this! Merged.

sillyousu commented:
>But for PostgreSQL same functionality might be not easy at all. Most interesting function (that contains tracing probes) exec_simple_query - is static.
Are there any limitations when using uprobe with static function?

goldshtn commented:
It needs to be a symbol visible in the binary so we can find its address and place a probe on it.

sillyousu commented:
gcc generate symbol for non-inlined static function [1] , so uprobe can work with `exec_simple_query` in some case (at least work for me, pg10+gcc5.4+default build config ).
But I don't know is it common or not.

[1] https://gcc.gnu.org/ml/gcc-help/2015-08/msg00142.html
